subsys/settings/src/settings_init.c

```diff
@@ -117,6 +117,52 @@ int settings_backend_init(void)
 
 	return rc;
 }
+
+#elif defined(CONFIG_SETTINGS_NVS)
+#include <device.h>
+#include <flash_map.h>
+#include "settings/settings_nvs.h"
+
+#define NVS_MAX_SECTOR_NUM (DT_FLASH_AREA_STORAGE_SIZE / \
+			    DT_FLASH_ERASE_BLOCK_SIZE)
+
+static struct settings_nvs default_settings_nvs;
+
+int settings_backend_init(void)
+{
+	int rc;
+	u32_t cnt = NVS_MAX_SECTOR_NUM;
+	struct flash_sector sectors[NVS_MAX_SECTOR_NUM];
+
+	rc = flash_area_get_sectors(DT_FLASH_AREA_STORAGE_ID, &cnt, sectors);
+	if (rc) {
```

Comment:
The default nvs system will take the complete storage area, so if a user would like to split up the storage area in two nvs systems (one for backup, one for runtime) this is not possible.

Reply:
It won't be an issue any more when DTS will let you define multiple storage areas for NVS.

Reply:
True, but you will have to do changes the settings_init.c again to reflect this.

My BAG (big ambitious goal) would be to try and separate any knowledge about the dts from the settings module. I would do this by defining items in Kconfig and then filling them up in the configuration stage.

So how to do this: 
a. Define a settings_nvs_default_storage_label (default equal to storage)
b. Define a settings_nvs_default_sector_count (default equal to 65535, meaning it will take the maximum available sectors)
c. Define a settings_nvs_default_sector_size_multiplier (default to 1)
(d. optionally you could also define settings_nvs_default_dev_name)

Using these defines and flashmap you can find out the size of the storage area and adapt (using the size_multiplier) sector_count to reflect the real maximum. There would be no reference to the dts at all (except in Kconfig). This gives clear separation between dts and the settings module, which is good.

Reply:
^^ sounds good. I would fetch settings_nvs_default_dev_name using API (less to configure).
I prefer to fulfill this in next patch.

Reply:
I am going to add the multiplier to the Kconfig. It seems to be a really useful configuration option for devices with very small erase blocks. The rest can be updated in the future patch as @nvlsianpu said.

Reply:
^^OK

Reply:
Maybe my proposal has been misunderstood, what I meant is define Kconfig variables:
`CONFIG_SETTINGS_NVS_DEFAULT_STORAGE_LABEL`, `CONFIG_SETTINGS_DEFAULT_SECTOR_COUNT` and `CONFIG_SETTINGS_DEFAULT_SECTOR_SIZE_MULTIPLIER` and use these to setup the default nvs backend. It is a very small change, I see no reason to delay this to a future patch.

Reply:
I see no value in defining CONFIG_SETTINGS_DEFAULT_SECTOR_COUNT. IMO, the number of sectors should cover the whole storage area that was assigned to the NVS backend. If the user wants to change the number of sectors within their application. they can override storage partition configuration using DTS overlays.

Reply:
It is impossible to transfer a label via Kconfig (I will welcome Kconfig enchantment for that). Kconfig is able to transfer strings as `CONFIG_<string-config> "a string"`.

Reply:
> I see no value in defining CONFIG_SETTINGS_DEFAULT_SECTOR_COUNT. IMO, the number of sectors should cover the whole storage area that was assigned to the NVS backend. If the user wants to change the number of sectors within their application. they can override storage partition configuration using DTS overlays.

@kapi-no, the storage area was never assigned to the NVS backend, the NVS backend is taking it without any reasoning. I can understand that it would default to this but not to do it without any intervention possibility from the user. If `CONFIG_SETTINGS_NVS_DEFAULT_SECTOR_COUNT` would be defaulted to a large value it would result in the same user experience (default it is set to the complete area) but it leaves the user the possibility to override this decision.

@nvlsianpu, so it would be possible to set `CONFIG_SETTINGS_NVS_DEFAULT_LABEL = "STORAGE"`, this would be fine. Then you could use this to get the info from flash_map or not ?

Reply:
^^ No. Flash areas are identified by labeled numbers.

Reply:
@nvlsianpu, this is bad. But hey, you know how I dislike flashmap, it does not provide any usability and is just taking up ram/flash.
To solve the problem we might be able to set the ID instead of the label, maybe using the trick from @carlescufi: `default $(dt_hex_val, DT_FLASH_AREA_STORAGE_ID)`

Reply:
^^ I know what you think about it.
> To solve the problem we might be able to set the ID instead of the label, maybe using the trick from @carlescufi: default $(dt_hex_val, DT_FLASH_AREA_STORAGE_ID)

Unfortunately this works only on project generation.

Reply:
OK, once again I have given enough alternatives. If you make the decision that "it is the right of the settings module to take all the storage area" be brave enough to communicate this decision in a warning when the settings subsystem is selected. My proposal would be "We as the designers of the settings module have decided that the storage area is ours. This area cannot be used for anything else."

Reply:
@Laczen Why not creating another storage area instead of that. It will be dedicated to the application.

Now we have one partition:
```
		/* Storage partition will be used by FCB/NFFS/NVS if enabled. */
		storage_partition: partition@1000 {
			label = "storage";
			reg = <0x00001000 0x00010000>;
		};
```
This can be replaced on demand, if needs. Bellow NVS partition is truncated, and flowed by partition of user storage. Which looks legible. This solution doesn't mix Kconfig and DTS properties in order to get bounduaries for NVS - I think that's good.
```
		/* Storage partition will be used by FCB/NFFS/NVS if enabled. */
		storage_partition: partition@1000 {
			label = "storage";
			reg = <0x00001000 0x00008000>;
		};
		/* Storage partition will be used by user application. */
		custom_storage_partition: partition@9000 {
			label = "user-storage";
			reg = <0x00009000 0x00007000>;
		};
```

Reply:
@nvlsianpu for two reasons:

1. DTS should be hardware only, at least as much as possible. If possible we should eliminate even the mcuboot, image0, image1, ... definitions. This is possible, using e.g. a Kconfig for flashmap. IMO it is not good to put even more software (logical volume) definitions in DTS
2. When changing the dts you need to do this in all board definitions -> large chance for errors. I know you can override these board definitions but you would still need to do that for all boards.

@nvlsianpu, I agree that you should not mix Kconfig and DTS properties to get/check the boundaries. By defining the label_id that will be used and using flash_map[label_id] for sizes, and flash_page_layout for page_size it is possible to set/check the boundaries without any DTS definition. To get the maximum sector count it is just incrementing a variable with the `sector_multiplier x page_size` until the size from flash_map[label_id] is reached.

Reply:
@Laczen 
comment to 1. I don't have strong preference on that where flash areas boundaries should be defined. Maybe worth to analyze how other environments deal with this (linux, Riot, MyNewt, mbed-os), sum-up this and work on that in dedicated issue. Lot of people will be interested in that. I feel that what you mention is a very good argument.
I don't know who have decided that flash areas are described via DTS - this idea might has its advocates. This will be <B>Switch to describe flash-partition via KConfig only</B> issue.
I'm keen to act on that if you want too. cc @hakonfam have you opinion on that.

comment to 2. Yes Kconfig is much less error prone. Honestly i was thinking about using override only when needed - not to changing any boards dts for that.

Reply:
@kapi-no Can you introduce CONFIG_SETTINGS_NVS_DEFAULT_SECTOR_COUNT back?

Reply:
yes

Reply:
done

Reply:
thanks